### PR TITLE
Fix build with Boost 1.77.0

### LINKED ISSUE #22482: Failure to build with Boost 1.77
Build testing the OpenBSD ports tree with Boost 1.77 beta1. bitcoin 0.21.1 fails to build as shown..

```
wallet/test/init_test_fixture.cpp:34:52: error: no member named 'BOOST_FILESYSTEM_C_STR' in 'boost::filesystem::path'
    std::ofstream f(m_walletdir_path_cases["file"].BOOST_FILESYSTEM_C_STR);
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
1 error generated.
```

### PR DESCRIPTION
**Pull request by sizeofvoid**
BOOST_FILESYSTEM_C_STR changed to accept the path as an argument.

### CODE REVIEW

**Comment by sizeofvoid** [comment]
This will fix https://github.com/bitcoin/bitcoin/issues/22482

**Comment by fanquake** [comment]
Concept ACK.

Even though we will hopefully be removing Boost Filesystem in master (#20744), fixing this is still needed, as we'll have to at least backport the change to `22.x` and `0.21`, as users building those branches being using newer Boost versions.

**Comment by MarcoFalke** [comment]
Looks like boost 1.77 was released with this, so Concept ACK. Didn't test with that boost version, though.

**Comment by laanwj** [comment]
Concept ACK, I don't have that boost version to test with yet.

**Comment by brad0** [comment]
> Concept ACK, I don't have that boost version to test with yet. Would be nice to see an independent confirmation that this solves the issue (maybe from @brad0 , the original reporter) before merging.

Yes, it is building fine now.

File: src/fs.cpp [hunk 1]
```diff
@@ -242,7 +242,11 @@ void ofstream::close()
 }
 #else // __GLIBCXX__
 
+#if BOOST_VERSION >= 107700
+static_assert(sizeof(BOOST_FILESYSTEM_C_STR(*fs::path())) == sizeof(wchar_t),
```
Comment (laanwj) [inline src/fs.cpp:246]: Wouldn't
```c++
static_assert(sizeof(*BOOST_FILESYSTEM_C_STR(fs::path())) == sizeof(wchar_t),
```
communicate the intent here better?

**Comment by hebasto** [comment]
Concept ACK.

File: src/fs.cpp (hunk 1 above)
Comment (hebasto) [reply src/fs.cpp:246]: Agree.

The [member access](https://en.cppreference.com/w/cpp/language/operator_member_access#Built-in_member_access_operators) operator has [higher precedence](https://en.cppreference.com/w/cpp/language/operator_precedence) than the [indirection operator](https://en.cppreference.com/w/cpp/language/operator_member_access#Built-in_indirection_operator).

**Review by hebasto — COMMENTED** [review]
Tested 5825876a2339be72f205cc850cb3b28dbb10c412 using boost 1.77 from [depends](https://github.com/bitcoin/bitcoin/issues/22482#issuecomment-902554036).

@laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/22713/files#r691185723) should be addressed before merging.

**Comment by hebasto** [comment]
@sizeofvoid 

Could you address [this](https://github.com/bitcoin/bitcoin/pull/22713#discussion_r691185723) comment to get this PR ready?

File: src/fs.cpp (hunk 1 above)
Comment (sizeofvoid) [reply src/fs.cpp:246]: Agree!

**Comment by hebasto** [comment]
@sizeofvoid 

Sorry, but it looks like you changed the wrong line in your recent push. No need to change the old `static_assert`. Only the new one should be fixed.

**Comment by sizeofvoid** [comment]
> @sizeofvoid
> 
> Sorry, but it looks like you changed the wrong line in your recent push. No need to change the old `static_assert`. Only the new one should be fixed.

Ops, sorry. Fixed!

**Comment by hebasto** [comment]
ACK acb7aad27ec8a184808aa7905887e3b2c5d54e9c

**Review by benthecarman — APPROVED** [review]
ACK acb7aad27ec8a184808aa7905887e3b2c5d54e9c

**Review by fanquake — APPROVED** [review]
ACK acb7aad27ec8a184808aa7905887e3b2c5d54e9c - tested the fix with Boost 1.77.0 and 1.71.0.

**Pull request merged**